src/Language/Spectacle/Checker/Step.hs

```diff
@@ -0,0 +1,177 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Language.Spectacle.Checker.Step
+  ( -- * Steps
+    Step (Step, _stepFrom, _stepTo),
+
+    -- ** Construction
+    makeStep,
+    makeReflexStep,
+    stepHash,
+    toStepImage,
+
+    -- ** Predicates
+    isReflexiveStep,
+
+    -- ** Lenses
+    stepFrom,
+    stepTo,
+
+    -- * Step Images
+    StepImage (StepImage, _stepImageFrom, _stepImageTo),
+
+    -- ** Lenses
+    stepImageFrom,
+    stepImageTo,
+
+    -- ** Operations
+    traceToSteps,
+  )
+where
+
+import Data.Hashable (Hashable (hash, hashWithSalt))
+import Data.Sequence (Seq, pattern (:<|))
+import qualified Data.Sequence as Seq
+import Lens.Micro (Lens', lens, (^.))
+
+import Data.Type.Rec (Rec)
+import Language.Spectacle.Checker.Fingerprint (Fingerprint (Fingerprint))
+import Language.Spectacle.Checker.World (World, worldFingerprint)
+
+-- ---------------------------------------------------------------------------------------------------------------------
+
+-- | The 'Step' data type is step given by some next-state relation where '_stepFrom' is the preceding world and
+-- '_stepTo' is the succeding world.
+--
+-- @since 0.1.0.0
+data Step spec = Step
+  { _stepFrom :: {-# UNPACK #-} !(World spec)
+  , _stepTo :: {-# UNPACK #-} !(World spec)
+  }
+  deriving (Eq)
+
+-- | @since 0.1.0.0
+instance Show (Rec spec) => Show (Step spec) where
+  show (Step from to) = show from ++ " --> " ++ show to
+  {-# INLINE show #-}
+
+-- | @since 0.1.0.0
+instance Hashable (Step spec) where
+  hash (Step from to) = stepHash (from ^. worldFingerprint) (to ^. worldFingerprint)
+  {-# INLINE hash #-}
+
+  hashWithSalt salt (Step from to) = stepHashWithSalt salt (from ^. worldFingerprint) (to ^. worldFingerprint)
+  {-# INLINE hashWithSalt #-}
+
+-- | Construct a 'Step' from a pair of worlds @𝑤@, @𝑢@ related by the 𝑅-step @𝑤 𝑅 𝑢@.
+--
+-- @since 0.1.0.0
+makeStep :: World spec -> World spec -> Step spec
+makeStep = Step
+{-# INLINE CONLIKE makeStep #-}
+
+-- | The function that implements 'hash' for 'Step's. Every 𝑅-step @𝑤 𝑅 𝑢@ can be converted to a hash uniquely
+-- identifying a step from @𝑤@ to @𝑢@ by joining the fingerprints for @𝑤@ to @𝑢@.
+--
+-- @since 0.1.0.0
+stepHash :: Fingerprint -> Fingerprint -> Int
+stepHash (Fingerprint fp) (Fingerprint fp') = hashWithSalt fp fp'
+{-# INLINE stepHash #-}
+
+-- | Forgets the concrete values in a 'Step' leaving only the fingerprints of the initial and terminal worlds in a
+-- 'StepImage'.
+--
+-- @since 0.1.0.0
+toStepImage :: Step ctx -> StepImage
+toStepImage (Step from to) = StepImage (from ^. worldFingerprint) (to ^. worldFingerprint)
+{-# INLINE toStepImage #-}
+
+-- | Similar to 'stepHash' but has an additional argument for adding salt to into the resulting hash.
+--
+-- @since 0.1.0.0
+stepHashWithSalt :: Int -> Fingerprint -> Fingerprint -> Int
```

Comment:
What is the reason for needing to salt fingerprint hashes?

Reply:
Spectacle doesn't require it necessarily but the minimal implementation for `Hashable` does.